### PR TITLE
[MUGEN] Add caching to MultiheadAttention

### PR DESCRIPTION
**Pull request by langong347**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* (to be filled)

### CODE REVIEW

File: test/modules/layers/test_attention.py [hunk 1]
```diff
@@ -17,164 +19,300 @@
...
+
+@pytest.fixture
+def kv(n_heads, input_shape, hidden_dim):
+    return torch.randn(1, n_heads, *input_shape, hidden_dim // n_heads)
+
+
+@pytest.fixture
+def full_attn(input_shape):
+    # TODO: retire causal once mask generation is moved out of FullAttention
+    #   causal inside FullAttention does not affect caching of k, v
```
Comment (RdoubleA) [inline test/modules/layers/test_attention.py:60]: nit: move this TODO comment outside of test and in FullAttention for easier tracking?

File: test/modules/layers/test_attention.py [hunk 2]
```diff
@@ -17,164 +19,300 @@
...
+            return MultiHeadAttention(
+                input_shape, hidden_dim, hidden_dim, n_heads, 1, causal, attn_module
+            )
+
+        return create_multihead_attn
+
+    def test_split_multihead(self, input_shape, multihead_attn, full_attn):
+        mha = multihead_attn(False, full_attn)
+        x = torch.randn(1, *input_shape, 6)  # (b, d1, ..., dn, c)
+        mha.n_head = 2
```
Comment (RdoubleA) [inline test/modules/layers/test_attention.py:82]: you could make `n_head` a parameter in `create_multihead_attn` so you can remove fixture and you don't have to modify it here

File: torchmultimodal/modules/layers/attention.py [hunk 3]
```diff
@@ -131,63 +137,50 @@ def __init__(
...
-
-        a = self.attn(q, k, v, decode_step, decode_idx)
+                    # append present k, v to past k, v
+                    # for autoregressive decoding inputs are flattened as 1D sequences
+                    # so are the cached tensors: (b, n_heads, seq_len, c)
+                    k_, v_ = self.cache["k"], self.cache["v"]
+                    self.cache["k"] = torch.cat([k_, k], dim=2)
+                    self.cache["v"] = torch.cat([v_, v], dim=2)
+                # override the present k, v with the cache
+                k, v = self.cache["k"], self.cache["v"]
```
[the remark this answers is not in the document]
Comment (langong347) [reply torchmultimodal/modules/layers/attention.py:181]: In the non-causal case, `k` and `v` will be on the start the entire sequence. We should just compute their projection once (this logic is missing in MUGEN) and cache them once to be used later.

In the causal case, `k` and `v` will be step-wise so the cache will be incremental.

File: torchmultimodal/modules/layers/attention.py [hunk 4]
```diff
@@ -224,17 +217,14 @@ def forward(
         q: Tensor,
         k: Tensor,
         v: Tensor,
-        decode_step: Optional[int] = None,
-        decode_idx: Optional[Iterable[int]] = None,
     ) -> Tensor:
         mask = torch.Tensor(self.mask) if self.causal else None
```
Comment (RdoubleA) [inline torchmultimodal/modules/layers/attention.py:221]: The mask is already created for you when you're decoding. what about the case when a user specifies an attention mask during encoding? I am thinking of making `FullAttention` the base self-attention class for all models. But in FLAVA, ALBEF there is attention masking used in the encoders. how do we reconcile this?

Comment (langong347) [reply torchmultimodal/modules/layers/attention.py:227]: By "mask created while decoding" Are you referring to 

https://github.com/facebookresearch/multimodal/blob/e274207e5c091d681bcb790d561fca0764148549/torchmultimodal/modules/layers/attention.py#L231-L232

?

I think mask should be created by a top level module (either vanilla transformer or that in the decoding mode) and passed in to the attention modules instead of being created inside the attention module which is not only inflexible but incurring the additional argument `shape`.

Is this consistent with "user specifies an attention mask during encoding"?

Comment (RdoubleA) [reply torchmultimodal/modules/layers/attention.py:227]: Yes I agree with that, I think it's better design and it's consistent with other models. FLAVA passes in the attention mask at the highest level at `FLAVAModel`'s forward, for example. I can move the mask out of the attention modules in a separate PR for unifying `SelfAttention` and `MultiHeadAttention`

**Review by RdoubleA — APPROVED** [review]

File: torchmultimodal/modules/layers/attention.py (hunk 4 above)
Comment (langong347) [reply torchmultimodal/modules/layers/attention.py:227]: Feel free to move the mask out of the attention module. I was thinking about doing that previously as the transformer decoder layers will depend on that.

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37865352).
